### ISSUE TITLE
General discussions of implementation and API details for various features.

### ISSUE THREAD

**Issue by ws909** [ws909/glfw#4]
This issue serves the simple purpose of preventing pollution of GLFW observers' notification inboxes. I'm sure they love our spam, but it can be distracting.

**Comment by ws909** [comment]
![bilde](https://user-images.githubusercontent.com/37029098/223210230-f261eb4f-bc3b-47c3-9e82-46f713aa48c6.png)
Signal with 9+ notifications.

**Comment by GamesTrap** [comment]
I now figured out GDI transparency with help from @GyrosGeier:
![image](https://user-images.githubusercontent.com/16561717/223279392-8927f73c-0794-4937-a6c7-f984c3106eda.png)

I also changed the error message for `_glfwSetWindowTaskbarBadgeStringWin32()`.

I will now implement a function to generate a discord like notification icon.

**Comment by GamesTrap** [comment]
> I will now implement a function to generate a discord like notification icon.

Added `GenerateGenericBadgeIcon()`
![image](https://user-images.githubusercontent.com/16561717/223293978-c1dce18a-7b57-4c04-84f3-ce1d097fc119.png)

**Comment by ws909** [comment]
I might be able to achieve the same on MacOS, but I’m not sure if I really see the usefulness in it, as there’s enough space for more letters. Also, a `glfwSetWindowBadge(NULL, 1, GLFW_BADGE_GENERIC)` can’t be relied upon on X11 and WL.

**Comment by GamesTrap** [comment]
So no generic badge then?

What should the API now look like?

Also what are your thoughts on the [taskbar-progress issue comment](https://github.com/glfw/glfw/pull/2286#issuecomment-1456418283)?

**Comment by ws909** [comment]
Well, a `GLFW_WIN32_GENERIC_BADGE` boolean window hint is what I first considered, and I think that’s the right solution. 

Didn’t we agree that this is a fine API?
(Just renaming)
```c
GLFWAPI void glfwSetWindowBadge(GLFWwindow* window, int count);

GLFWAPI void glfwSetWindowBadgeString(GLFWwindow* window, const char* string); 
```

> Also what are your thoughts on the https://github.com/glfw/glfw/pull/2286#issuecomment-1456418283?

Yeah, that’s great. I can make those changes later today.

**Comment by GamesTrap** [comment]
> Didn’t we agree that this is a fine API?
> (Just renaming)

Yep, changed in https://github.com/GamesTrap/glfw/commit/367a50e82d7333c655953fae3706877423c4bc5f

> Well, a `GLFW_WIN32_GENERIC_BADGE` boolean window hint is what I first considered, and I think that’s the right solution.

What should happen when enabling the `GLFW_WIN32_GENERIC_BADGE` window hint?

Should a call to `glfwSetWindowBadge()` enable the generic badge and ignore the badge count?

**Comment by ws909** [comment]
Yeah, isn’t that what makes sense? For simplicity, callers can be required to set the window hint prior to activating the badge, for it to have an effect.

**Comment by GamesTrap** [comment]
I've now added  the generic badge window hint https://github.com/GamesTrap/glfw/commit/de81fae2688edc9390c19c78da983e9b4bb879b9

**Comment by ws909** [comment]
If we can consider this done now, you should probably merge the finalized version of `taskbar-progress` into `application-badge`, and create a PR for it.

**Comment by ws909** [comment]
I'm updating `glfwSetWindowIcon` to set the Dock icon on MacOS, but I'm expecting the implementation to break when `glfwSetWindowBadge` or `glfwSetWindowProgressIndicator` are called, so I'm considering just building on top of the application-badge branch instead of from scratch, like my branch is now.

Edit: copied the code I suspected would break the icon, and tested with it, and it ran fine, so it's ok. No need to build on top of the other branches.

**Comment by GamesTrap** [comment]
> If we can consider this done now, you should probably merge the finalized version of `taskbar-progress` into `application-badge`, and create a PR for it.

Will do this in the next couple of days.
Sorry for not responding, I got ill a couple of days ago and am still on my way to fully recover from it.

**Comment by ws909** [comment]
Don’t worry about that. Take care. :)

**Comment by ws909** [comment]
Just postpone this until you’re no longer sick. I’ll take a closer look at it again too.

**Comment by GamesTrap** [comment]
> merge the finalized version of `taskbar-progress` into `application-badge`

Done.

> create a PR for it

~~Will do in a few hours.~~ Postponed :D

**Comment by ws909** [comment]
I just had a quick overview, and I see some things missing, so just delay the PR for now. I’ll make some additions within the next few days. Camilla hasn’t given any attention to GLFW the last days either, so we’re in no hurry.

**Comment by ws909** [comment]
It'll have to be this week. I haven't had time to do it yet.

**Comment by GamesTrap** [comment]
Are there any issues left for the application badge branch or can I create the upstream PR now?

**Comment by ws909** [comment]
I think you can open a PR now. I can’t remember if I had any other changes in mind.

**Comment by GamesTrap** [comment]
Opened PR:
- https://github.com/glfw/glfw/pull/2312

**Comment by ws909** [comment]
:) Don't worry. I get e-mails when you open PRs in GLFW's repository. :P

**Comment by GamesTrap** [comment]
I've worked on the DBus implementation a bit to get rid of the repetition.

New functions I've added so far:
```c
dbus_bool_t _glfwNewMessageSignalDBusPOSIX(const char* objectPath, const char* interfaceName, const char* signalName, struct DBusMessage** outMessage);
dbus_bool_t _glfwOpenContainerDBusPOSIX(struct DBusMessageIter* iterator, int DBusType, const char* signature, struct DBusMessageIter* subIterator);
dbus_bool_t _glfwCloseContainerDBusPOSIX(struct DBusMessageIter* iterator, struct DBusMessageIter* subIterator);
dbus_bool_t _glfwAppendDataDBusPOSIX(struct DBusMessageIter* iterator, int DBusType, const void* data);
dbus_bool_t _glfwAppendDictDataDBusPOSIX(struct DBusMessageIter* iterator, int keyType, const void* keyData, int valueType, const void* valueData);
dbus_bool_t _glfwSendMessageDBusPOSIX(struct DBusMessage* message);
```

```c
void _glfwUpdateBadgeDBusPOSIX(dbus_bool_t badgeVisible, int badgeCount)
{
    struct DBusMessage* msg = NULL;

    if(!_glfw.dbus.handle || !_glfw.dbus.connection || !_glfw.dbus.desktopFilePath)
        return;

    long long badgeCountLL = badgeCount;

    //Signal signature:
    //signal com.canonical.Unity.LauncherEntry.Update (in s app_uri, in a{sv} properties)

    struct DBusMessageIter args;
    memset(&args, 0, sizeof(args));

    if(!_glfwNewMessageSignalDBusPOSIX("/org/glfw", "com.canonical.Unity.LauncherEntry", "Update", &msg))
        return;

    dbus_message_iter_init_append(msg, &args);

    //Setup app_uri parameter
    _glfwAppendDataDBusPOSIX(&args, DBUS_TYPE_STRING, &_glfw.dbus.desktopFilePath);

    //Set properties parameter
    struct DBusMessageIter sub1;
    memset(&sub1, 0, sizeof(sub1));

    _glfwOpenContainerDBusPOSIX(&args, DBUS_TYPE_ARRAY, "{sv}", &sub1);

    //Set count visible property
    const char* countVisibleStr = "count-visible";
    _glfwAppendDictDataDBusPOSIX(&sub1, DBUS_TYPE_STRING, &countVisibleStr, DBUS_TYPE_BOOLEAN, &badgeVisible);

    //Set count value property
    const char* countStr = "count";
    _glfwAppendDictDataDBusPOSIX(&sub1, DBUS_TYPE_STRING, &countStr, DBUS_TYPE_INT64, &badgeCountLL);

    _glfwCloseContainerDBusPOSIX(&args, &sub1);

    _glfwSendMessageDBusPOSIX(msg);

    //Free the message
    dbus_message_unref(msg);
}
```

Any suggestions?

**Comment by ws909** [comment]
I haven't read the DBus specification myself, so don't count me in on the knowledge of the API itself, but I merged the _application-badge_ branch into my _notifications_ branch to better understand how the FreeDesktop Notifications API can be layered on top of DBus. That's what led me to those comments on the PR.

So, for suggestions:
If I designed the way GLFW works with DBus, I'd probably design them as layers, where GLFW sits on top of DBus, so I'd strip all references to GLFW from _posix_dbus.c_, and pass everything as parameters instead. Or, even better, wrap this in a class, with the recurring arguments as fields of the class. There's likely nothing gained from going to the lengths of emulating a class in C for this purpose.

Other than that, it looks good to me. It seems similar to what I saw while looking at Libnotify's source code.

I realized there's multiple things that's needed to make the Notifications API work. For instance, GLFW needs to receive signals over the bus. Can these be read from the bus in `glfwPollEvents` and `glfwWaitEvents`? Would there be any issues with `glfwWaitEvents` specifically? 

Btw: `_glfwNewMessageSignalDBusPOSIX("/org/glfw"`; Is this constrained to the application, or can any GLFW application intercept this? Any issues with multiple GLFW applications communicating over DBus at once?

**Comment by GamesTrap** [comment]
> For instance, GLFW needs to receive signals over the bus. Can these be read from the bus in `glfwPollEvents` and `glfwWaitEvents`? Would there be any issues with `glfwWaitEvents` specifically?

Here is a C++ example on how to listen for messages/signals:
https://dbus-cxx.github.io/signal_receiver_c_8cpp-example.html

**Comment by GamesTrap** [comment]
> Btw: `_glfwNewMessageSignalDBusPOSIX("/org/glfw"`; Is this constrained to the application, or can any GLFW application intercept this? Any issues with multiple GLFW applications communicating over DBus at once?

That's an oversight on my part, the object path should be unique for each program and even each instance of a program.

I think it would make sense to add the process id to the object path to ensure it is unique for each running instance/program using glfw.

Example: "/org/glfw/24418"

**Comment by ws909** [comment]
> That's an oversight on my part, the object path should be unique for each program and even each instance of a program.

That must be fixed for both of the PRs; badge and progress.

**Comment by ws909** [comment]
> I think it would make sense to add the process id to the object path to ensure it is unique for each running instance/program using glfw.

As long as that’s the intended usage for it; if not, the application’s name should be used. If this path is used to identify the application between executions, it shouldn’t change.

**Comment by GamesTrap** [comment]
> As long as that’s the intended usage for it; if not, the application’s name should be used. If this path is used to identify the application between executions, it shouldn’t change.

The application name would be the same for two concurrently running instances of the same program so it isn't unique in that case. Linux `getpid()` function also states `getpid() returns the process ID (PID) of the calling process.
       (This is often used by routines that generate unique temporary
       filenames.)`

**Comment by GamesTrap** [comment]
```c
        //Request name

        //Get name of the running executable
        char exeName[PATH_MAX];
        memset(exeName, 0, sizeof(char) * PATH_MAX);
        if(readlink("/proc/self/exe", exeName, PATH_MAX) == -1)
        {
            _glfwInputError(GLFW_PLATFORM_ERROR, "Failed to get name of the running executable");
            return;
        }
        char* exeNameEnd = strchr(exeName, '\0');
        char* lastFound = strrchr(exeName, '/');
        if(!lastFound || !exeNameEnd)
        {
            _glfwInputError(GLFW_PLATFORM_ERROR, "Failed to get name of the running executable");
            return;
        }
        unsigned int exeNameLength = (exeNameEnd - lastFound) - 1;

        //"org.glfw.<exe_name>_<pid>"
        char* objectPath = _glfw_calloc(21 + exeNameLength, sizeof(char));
        memset(objectPath, '\0', 21 + exeNameLength * sizeof(char));

        const pid_t pid = getpid();
        sprintf(objectPath, "org.glfw.%s_%d", lastFound + 1, pid);

        printf("The object path is: %s\n", objectPath);
        const int res = dbus_bus_request_name(_glfw.dbus.connection, objectPath, DBUS_NAME_FLAG_REPLACE_EXISTING, &_glfw.dbus.error);
```

`The object path is: org.glfw.window_34289`

Good?

**Comment by GamesTrap** [comment]
Added the unique signal/bus names:

https://gist.github.com/GamesTrap/4deb10a29867aae0c72d8359fec78958

I will add the other changes from before into this, then push this into taskbar-progress and lastly merge the taskbar-progress branch into application-badge.

**Comment by ws909** [comment]
Looks good to me. :)

**Comment by GamesTrap** [comment]
Updated PRs, had to fix DBus crashing because of illegal characters in the executable name (everything that is not alnum is now ignored)

**Comment by ws909** [comment]
Would you by any chance now what happens when a notification is clicked or dismissed on Linux, and the application is not running? On MacOS, the application is opened, and the UNUserNotificiationCenterDelegate’s handler method is called. The delegate is set at app init, so Cocoa calls this method before the applicationDidFinishLaunching method is called. I don’t know how the system passes the notification event to the application at startup; possibly over a socket or as cli arguments, or a combination. I’d like to support this on Linux as well, but I don’t know if it’s possible. Responding to these events is possible on Linux, if the application is running, but I don’t know what it takes to do it if it’s not running.

**Comment by GamesTrap** [comment]
The [spec](https://specifications.freedesktop.org/notification-spec/latest/ar01s09.html) has an `ActionInvoked` signal but it can only be processed/listened to while the program is running.

> Responding to these events is possible on Linux, if the application is running, but I don’t know what it takes to do it if it’s not running.

I think you would have to use a background service/program that listens to the signal in order to do this like macOS.

I'm not an expert on DBus stuff so there may be other ways to do what you want.

**Comment by ws909** [comment]
> The [spec](https://specifications.freedesktop.org/notification-spec/latest/ar01s09.html) has an ActionInvoked signal but it can only be processed/listened to while the program is running.

Yeah, that I’m aware of, but again, while running. 

> I think you would have to use a background service/program that listens to the signal in order to do this like macOS.

That’s a very big drawback. I wonder why they’d design it this way. Clicking on notifications on Windows also opens the application. The FreeDesktop Notifications specification/API is old, but I’d assume they’d have updated it by now, or at least have added some paragraphs about this there. There was nothing more listed under capabilities and hints. Another feature lacking on Linux, is also the possibility of associating data with notifications, and retrieve it from the server. The workaround is out of scope for GLFW (I believe). But, the more I read of the documentation for notifications on Apple platforms, the more sure I am that the framework handles this itself in the same way GLFW would have to: cache notification contents in the application’s own caches. 

Knowing how notifications work on Linux; if the application is opened without having a background service, would be immensely useful. I guess I’ll have to search more.
